builtin/union_dataset.cc

```diff
@@ -0,0 +1,397 @@
+#include "mldb/types/any_impl.h"
+#include "mldb/types/structure_description.h"
+#include "mldb/types/vector_description.h"
+
+using namespace std;
+
+
+namespace MLDB {
+
+
+/*****************************************************************************/
+/* UNION DATASET CONFIG                                                      */
+/*****************************************************************************/
+
+DEFINE_STRUCTURE_DESCRIPTION(UnionDatasetConfig);
+
+UnionDatasetConfigDescription::
+UnionDatasetConfigDescription()
+{
+    nullAccepted = true;
+
+    addField("datasets", &UnionDatasetConfig::datasets,
+             "Datasets to unify together");
+}
+
+static RegisterDatasetType<UnionDataset, UnionDatasetConfig>
+regUnion(builtinPackage(),
+          "union",
+          "Unify together several datasets",
+          "datasets/UnionDataset.md.html");
+
+extern std::shared_ptr<Dataset> (*createUnionDatasetFn) (MldbServer *, vector<std::shared_ptr<Dataset> > datasets);
+
+std::shared_ptr<Dataset> createUnionDataset(
+    MldbServer * server, vector<std::shared_ptr<Dataset> > datasets)
+{
+    return std::make_shared<UnionDataset>(server, datasets);
+}
+
+struct UnionDataset::Itl
+    : public MatrixView, public ColumnIndex {
+
+    ML::Lightweight_Hash<RowHash, pair<int, RowHash> > rowIndex;
+
+    // Datasets that it was constructed with
+    vector<std::shared_ptr<Dataset> > datasets;
+
+    Itl(MldbServer * server, vector<std::shared_ptr<Dataset> > datasets) {
+        if (datasets.empty()) {
+            throw ML::Exception("Attempt to unify no datasets together");
+        }
+        this->datasets = datasets;
+        int indexWidth = getIndexBinaryWidth();
+        if (indexWidth > 31) {
+            throw ML::Exception("Too many datasets in the union");
+        }
+        for (int i = 0; i < datasets.size(); ++i) {
+            for (const auto & rowName: datasets[i]->getMatrixView()->getRowNames()) {
+                rowIndex[RowHash(PathElement(i) + rowName)] =
+                    make_pair(i, RowHash(rowName));
+            }
+        }
+    }
+
+    int getIndexBinaryWidth() const {
+        return ceil(log(datasets.size()) / log(2));
+    }
+
+    int getIdxFromRowName(const RowName & rowName) const {
+        // Returns idx > -1 if the index is valid, -1 otherwise
+        if (rowName.size() < 2) {
+            return false;
+        }
+        string idxStr = (*(rowName.begin())).toUtf8String().rawString();
+        int idx = atoi(idxStr.c_str());
+        if (idx == 0 && idxStr != "0") {
+            // atoi returns 0 on error, make sure it's a real 0
+            return -1;
+        }
+        if (idx > datasets.size()) {
+            return -1;
+        }
+        return idx;
+    }
+
+    struct UnionRowStream : public RowStream {
+
+        UnionRowStream(const UnionDataset::Itl* source) : source(source)
+        {
+            cerr << "UNIMPLEMENTED " << __FILE__ << ":" << __LINE__ << endl;
```

Comment:
Or at the very least, it should throw that it's unimplemneted not silently do the wrong thing.